### ISSUE TITLE
macOS template

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#1982]
https://groups.google.com/d/msg/qubes-users/RiVntUzgJmY/rXMtXD3WKQAJ

**Comment by lucaszanella** [comment]
I couldn't find information regarding Qubes 4.1. I suppose those pacthes were added already. Do you have any new information?

**Comment by andrewdavidwong** [comment]
> I couldn't find information regarding Qubes 4.1. I suppose those pacthes were added already. Do you have any new information?

No new info on this, as far as I know.

**Comment by fabiogermann** [comment]
I also couldn't find a pull request in  QubesOS/qubes-vmm-xen for those patch files - I assume that would be the appropriate repo. A first step could be to get them into the standard build.

**Comment by lucaszanella** [comment]
I don't understand this step

"Update the attached files with the MAC address and IP address assigned to your appvm"

I don't know where to find the MAC and IP addresses in my newly created HVM.

I also don't know if my QEMU is recent enough, and I don't even know which qemu I have. The qemu's that appear in my dom0 are qemu-img-xen and qemu-ndb-xen, for which I can't find the version.

I tried generating an image from https://github.com/kholia/OSX-KVM and booting into an HVM to see if the patches were working but it won't even boot. Maybe the pen drive method from the qubes forum post is still needed. However I'm trying to do an auditable macOS VM so I don't like the idea of starting with a pkg.

If someone finds this post and has a cleaner and auditable way of doing all this, please talk to me :)

**Comment by ideologysec** [comment]
Different/better strange news (separate from the difficulties of trying to map old Xen -> new macOS instructions and the Xen!=KVM apportionment of effort):
https://communities.vmware.com/ideas/1803

Apple dev linked documentation: https://developer.apple.com/documentation/paravirtualizedgraphics

**Comment by lucaszanella** [comment]
@ideologysec 

Very nice, I just hope Qubes gains ARM support otherwise we'll not have much years to enjoy macOS on it.

I was trying to experiment with UEFI boot: https://github.com/QubesOS/qubes-issues/issues/5767 to try to boot https://github.com/foxlet/macOS-Simple-KVM which runs great on Qemu but I did not go too far yet.

**Comment by tbrtje** [comment]
Is this currently being worked on? Everything I could find was for Qubes 3.
I cannot get it to work on Qubes 4.

**Comment by ideologysec** [comment]
I have doubts this will ever be much of a priority, especially as Apple moves away from Intel. Shame, though.

I'm about to wipe my laptop and try this, though.

**Comment by eduncan911** [comment]
Funny... I was JUST watching this now 5 minutes ago and thought of exactly THIS issue to rehash!

https://youtube.com/watch?v=ATnpEOo3GJA

The GitHub uses KVM, but the key VM info is there that could be copied over to a Xen VM/template.

Their KVM setup uses a dedicated GPU and a dedicated PCIe cars for USB (keyboard, mouse, etc).

It's a serious start, knowing the boot bars now.

I don't have a qubes machine at this time or I'd work on it

**Comment by lucaszanella** [comment]
@eduncan911 the github project that ables to run macOS on KVM uses KVM + Qemu. Qubes uses Xen + Qemu. The main reason macOS wont run on Qubes easily is the fact that all VMs in Qubes boot in BIOS mode, and macOS uses UEFI. I've learned that Xen supports UEFI and the Xen currently built into Qubes already supports it too as an experimental feature. I even discovered how to enable it but I couldn't boot windows or linux in UEFI mode. Didn't try anything more but if we make UEFI boot work then it should be relatively easy to make macOS run...

**Comment by tbrtje** [comment]
Uefi isn't required actually. You can create an opencore installation that runs on a standard bios, but I have no idea how to do this with qubes, because I don't really understand how I can attach an existing qcow2 image in qubes.

**Comment by lucaszanella** [comment]
@tbrtje yes, the old tutorial does that way, but I think that with UEFI the KVM project will run with much fewer modifications

**Comment by agowa338** [comment]
also for windows vms the uefi is more or less a must by now. It just boots way faster and from experience also has less quirks with updates...

**Comment by ideologysec** [comment]
So, it is possible to run macOS as a guest with no Metal support - VMware Fusion on macOS runs macOS guests prior to Big Sur unaccelerated; performance is just crap.

Big Sur adds support for guest accelerated graphics, and even prior macOS versions started supporting virtual drivers:

https://www.kraxel.org/blog/2019/06/macos-qemu-guest/

**Comment by lucaszanella** [comment]
@ideologysec does it mean that it's possible to run macOS on KVM with accelerated graphics? How could I do that? I have a linux machine with macOS on KVM and it runs slow because of video performance. Couldn't make an AMD graphics card work in PCI passtrhough.

**Comment by ninavizz** [comment]
LOL @DemiMarie you have figured me out! Yes, a big audible gasp happened once I read this. That said, MacOS is a biggy GPU/CPU/RAM/etc hog... and the graphics programs I work with, are no different. Adobe apps, especially. Still mighty jazzed to see this may happen at some point, though! :)

**Comment by DemiMarie** [comment]
> LOL @DemiMarie you have figured me out! Yes, a big audible gasp happened once I read this.

I laughed when I saw your reply :rofl:.

> That said, MacOS is a biggy GPU/CPU/RAM/etc hog... and the graphics programs I work with, are no different. Adobe apps, especially. Still mighty jazzed to see this may happen at some point, though! :)

Not surprised.  Qubes OS works just fine with most of its CPU and RAM used by a single qube, though.  And I am almost certain that the most powerful Mac has at least twice the CPU and RAM of the least powerful one.

The GPU is going to be the hard part.  I do plan on adding GPU acceleration to Qubes OS at some point in the future, but I currently plan on running a custom userspace driver in `sys-gui` and a matching frontend driver in the qubes.  That way, I can ensure that everything is written with security and minimal attack surface in mind, which is what I would need to be comfortable turning it on by default.  This won’t be compatible with macOS or Windows, as I have no interest in writing drivers for either of these.  Additionally, this will almost certainly be less heavily optimized than the official drivers ― for example, I may choose to have the driver be single-threaded rather than try to deal with synchronization.  Finally, this will only support hardware that comes with complete documentation, which currently means Intel and AMD hardware.

For graphics-intensive applications, external GPUs are likely the way forward.  Qubes OS does not currently support PCIe hotplug, but this will be fixed.

**Comment by agowa** [comment]
No, not really. However instead of macOS you could try https://ravynos.com/. This is an opensource alternative that claims to be compatible with most of the macOS software.
I still have checking it out more closely one on my ToDo list. However as I don't know when I'll actually have time to do so feel free to give it a shot and report back. Integrating it is probably also way easier than macOS itself.

**Comment by v6ak** [comment]
> This is an opensource alternative that claims to be compatible with most of the macOS software.

It doesn't. Quoting from https://ravynos.com/faq :


> Can I run macOS apps on ravynOS?
>
>    No - not yet. This is a major goal and we're working towards it. So far, some trivial "Hello World" binaries in C and Objective-C have been compiled on macOS Big Sur and the resulting Mach-O binaries ran on ravynOS using our in-dev Hybrid linker without any Frameworks or tools from macOS. Source code of some simple Cocoa AppKit-based applications have been built on ravynOS and run as native applications (e.g. Install ravynOS.app). While this is all very exciting, you will have to wait a bit longer to run Photoshop or Xcode.

**Comment by v6ak** [comment]
> It seems that OpenCore doesn't support Xen right off the bat, and there are some implementations that need to be addressed.

I've grepped xen in OpenCore's sources. There are few occurrences, probably added shortly after your comment. There is the relevant changelog entry: commit dce5a8ff8 by Vitaly Cheptsov at 2024-01-21 16:47:20 +0300:

> Added TSC frequency calculation for xen hypervisor, thx @netanelc305

Not sure if this was your issue, but it looks like they are aware of Xen. I haven't tried running macOS on Qubes, though.

**Comment by agowa** [comment]
> > This is an opensource alternative that claims to be compatible with most of the macOS software.
> 
> It doesn't. Quoting from https://ravynos.com/faq :

Oh, I was referring to the projects goal once it is out of it's current very early alpha state. The FAQ is talking about the current state. Sorry for not making it more clear.

**Comment by DemiMarie** [comment]
Given that the future of macOS on x86 is limited at best, I’m tempted to close this as a duplicate of #7476 (AArch64 support), which would allow running Qubes OS on Apple silicon directly.  Eventually, I expect Apple to drop support for x86 altogether, so at least I am not too interested in supporting macOS on an architecture that is a dead end for it.

**Comment by agowa** [comment]
@DemiMarie It's not really a duplicate as one is for running macOS VMs on QubesOS and the other one is for running QubesOS on AArch64 Macbooks.

However if I recall correctly apple only ever provided licenses for using MacOS on Mac hardware, so I think I can understand where you're coming from.

**Comment by yangm97** [comment]
> Given that the future of macOS on x86 is limited at best [...]

Well yeah, that's why running Qubes OS on so called "vintage" macs, along with x86_64 macOS guests, makes sense.

![Image](https://github.com/user-attachments/assets/62c5c7ad-3c3d-4b26-9b0c-78aeb30bab7d)

s/children/environment

According to Apple, the 2017 MBP I'm typing this from becomes eWaste material in 12 days, meaning they are going to stop issuing security updates to macOS Sonoma.

So, shouldn't an aftermarket OS support aftermarket hardware? Between T1/T2 and M1 users, I believe users of the former are much more likely to be looking for an alternative OS _right now_ than the latter. 

> I expect Apple to drop support for x86 altogether, so at least I am not too interested in supporting macOS on an architecture that is a dead end for it.

There's also a case to be made that Rosetta won't be around forever, so if you consider emulating Rosetta is more complex than running a x86_64 macOS under qemu, wouldn't it be better to just skip the middleman for the sake of software preservation?

I believe I'm not alone when I say that I would enjoy running Qubes on my sad slab of aluminum eWaste, but the apparent lack of interest on supporting macOS guests unfortunately makes running Linux much more appealing instead.

**Comment by DemiMarie** [comment]
The problem is that macOS on x86_64 is itself of limited viability.

If you plan to run x86_64 macOS in a VM with no network access, that might be okay.

**Comment by andrewdavidwong** [comment]
@yangm97: This issue is about running macOS inside of Qubes OS, not about running Qubes OS on Mac hardware.

**Comment by yangm97** [comment]
@andrewdavidwong I am making a point that supporting x86_64 macOS guests are needed because Intel macs aren't going to disappear overnight.

**Comment by andrewdavidwong** [comment]
> [@andrewdavidwong](https://github.com/andrewdavidwong) I am making a point that supporting x86_64 macOS guests are needed because Intel macs aren't going to disappear overnight.

1. The prevalence of Intel Macs is irrelevant to Qubes OS. Supporting old Mac hardware isn't a goal of the project or part of its mission, just a nice bonus if it happens to work out. Qubes is a security-oriented operating system, not an obsolete-hardware-oriented operating system. If you want someone to blame, then blame Apple for abandoning their own products. Don't blame a tiny FOSS project for not bailing out a $4T multinational corportaion. The Qubes OS Project's only sin here is that it happens to exist in the same world as Apple at the same point in history.

2. It doesn't follow logically that macOS guests are needed in Qubes OS running on Intel Macs. Just because you have an Intel Mac doesn't mean you *have* to run macOS guests on it. You could instead run Fedora, Debian, or whichever other supported guests you like.

**Comment by DemiMarie** [comment]
I don’t work for ITL anymore, but I **strongly** recommend closing this issue as “won’t fix”.

Intel Macs and x86_64 macOS are obsolete, and Qubes OS does not support Arm64.  Supporting Arm64 would be an interesting project, but it would require a lot of work in Xen.  For instance, device passthrough on Arm64 isn’t security supported right now, and Xen does not support PCI on Arm64 at all yet.

Furthermore, the macOS EULA only allows running macOS on Apple hardware.  Supporting generic Arm64 platforms (to the extent they exist outside of servers) would not provide Apple silicon support.  Apple silicon is simply too different.  For instance, it has nonstandard IOMMUs that Xen would need to be aware of.  To the best of my knowledge, nobody else is interested in running Xen on Apple silicon, so ITL would have the entire maintenance burden.  Since some of this burden relates to security, not all of it could be handled by the community.

Instead of trying to support Apple silicon with Xen, a much better use of resources would be to port Qubes OS to KVM.  This would provide other benefits, such as better power management (on all platforms) and Arm64 support.  I doubt that the userspace VMM Qubes OS would use would support emulating the devices macOS needs, but it would be possible to run QEMU + KVM under nested virtualization.  If QEMU supports macOS guests, macOS could then be run as a nested QEMU guest on an Apple silicon host.

**Comment by marmarek** [comment]
I agree, support for obsolete hardware/software (which x86 Mac OS is at this point) is not the goal of _this_ project.

**Comment by andrewdavidwong** [comment]
> Intel Macs are obsolete and Qubes OS does not support Arm64. [...]

> I agree, support for obsolete hardware/software (which x86 Mac OS is at this point) is not the goal of this project.

Maybe I'm missing something, but I don't see how these comments are relevant to this issue, since this issue is about having a macOS template in Qubes OS, not about running Qubes OS on Apple hardware.

**Comment by DemiMarie** [comment]
> > Intel Macs are obsolete and Qubes OS does not support Arm64. [...]
> 
> > I agree, support for obsolete hardware/software (which x86 Mac OS is at this point) is not the goal of this project.
> 
> Maybe I'm missing something, but I don't see how these comments are relevant to this issue, since this issue is about having a macOS template in Qubes OS, not about running Qubes OS on Apple hardware.

Qubes OS does not support arm64, so it can’t run arm64 macOS in a VM on any hardware.  Also, the macOS EULA only allows running macOS on a Mac, so a macOS template only makes sense if Qubes OS supports Apple hardware.  I edited my comment to clarify the matter.

I _do_ think that working with Xen upstream on arm64 support makes sense.  There are enough arm64 server systems that ITL would not have to do all of the maintenance on its own.  Apple silicon is just too nonstandard.

**Comment by agowa** [comment]
@DemiMarie not even VMware was honoring that part of the EULA though. Ok they officially did but in practice they never did. All that needed to be done was changing some config values within the advanced properties...

Besides that I agree that focusing on Xen first would be more meaningful for now.

**Comment by DemiMarie** [comment]
@agowa VMware isn’t the one who would be liable for violations.  The user would be.  A better way to understand this is to look at the providers of macOS CI services.  They all use real Apple hardware.

**Comment by agowa** [comment]
@DemiMarie Well officially they at least pretended to enforce this part of apples EULA. Where as inofficially you just had to change the value of a (badly documented) advanced property.
